File: Java/n-queens-ii.java
//  N-Queens II

// Time O(N!), Space O(1)
class Solution {
    private int ans = 0;
    
    public int totalNQueens(int n) {
        int[] C = new int[n];
        Arrays.fill(C, -1);
        dfs(n, 0, C);
        return this.ans;
    }
    
    private void dfs(int n, int row, int[] C) {
        if (row == n) {
            this.ans ++;
            return;
        }
        for (int col=0; col<n; col ++) {
            if (!isValid(C, row, col))
                continue;
            C[row] = col;
            dfs(n, row+1, C);
            C[row] = -1;
        }
    }
    
    private Boolean isValid(int[] C, int row, int col) {
        for (int i=0; i<row; i++) 
            if (col==C[i] || Math.abs(i-row)==Math.abs(C[i]-col))
                return false;
        return true;
    }
}

File: leetcode/n-queens-ii.cc
//  N-Queens II

class Solution {
public:
    int totalNQueens(int n) {
        int res = 0;
        vector<int> flag(n, -1);
        dfs(n, 0, flag, res);
        return res;
    }
    
    void dfs(int n, int cur, vector<int> &flag, int &res) {
        if (cur == n) {
            res ++;
            return;
        }
        for (int col=0; col<n; col++) {
            if (!is_valid(cur, col, flag))  continue;
            flag[cur] = col;
            dfs(n, cur+1, flag, res);
            flag[cur] = -1;
        }
    }
    
    bool is_valid(int row, int col, vector<int> &flag) {
        for (int i=0; i<row; i++) 
            if (col==flag[i] || abs(i-row)==abs(flag[i]-col))
                return false;
        return true;
    }
};